src/main/java/net/schmizz/sshj/common/CircularBuffer.java

```diff
@@ -0,0 +1,151 @@
+    private int rpos;
+    private int wpos;
+
+    public CircularBuffer(int size, int maxSize) {
+        this.maxSize = maxSize;
+        int initialSize = getNextSize(size);
+        if (initialSize > maxSize) {
+            throw new IllegalArgumentException(
+                String.format("Initial power of 2 size %d larger than maximum size %d", initialSize, maxSize));
+        }
+        this.data = new byte[initialSize];
+        this.rpos = 0;
+        this.wpos = 0;
+    }
+
+    public int available() {
+        int available = wpos - rpos;
+        return available >= 0 ? available : available + data.length; // adjust if wpos is left of rpos
+    }
+
+    public int maxPossibleRemainingCapacity() {
+        // Get remaining in current buffer.
+        int remaining = rpos - wpos;
+        if (remaining < 0) {
+            remaining += data.length; // adjust if rpos is left of wpos
+        }
+        // Add the maximum amount it can grow.
+        return remaining + maxSize - data.length;
+    }
+
+    private void ensureAvailable(int a)
+            throws CircularBufferException {
+        if (available() < a) {
+            throw new CircularBufferException("Underflow");
+        }
+    }
+
+    void ensureCapacity(int capacity) {
+        int available = available();
+        int remaining = data.length - available;
+        // If capacity fits exactly in remaining space, expand it; otherwise, wpos will equal rpos after writing.
+        if (remaining <= capacity) {
+            int cw = available + capacity + 1;
+            byte[] tmp = new byte[getNextSize(cw)];
+            // Copy the one or two segments between rpos and wpos and start again from 0.
+            if (wpos >= rpos) {
+                System.arraycopy(data, rpos, tmp, 0, available);
+                wpos -= rpos; // move wpos left so that rpos starts from 0
+            } else {
+                int tail = data.length - rpos;
+                System.arraycopy(data, rpos, tmp, 0, tail); // segment right of rpos
+                System.arraycopy(data, 0, tmp, tail, wpos); // segment left of wpos
+                wpos += tail; // move wpos right to let rpos wrap around to 0
+            }
+            rpos = 0;
+            data = tmp;
+        }
+    }
+
+    public void readRawBytes(byte[] buf, int off, int len)
+            throws CircularBufferException {
+        ensureAvailable(len);
+
+        int rposNext = rpos + len;
+        if (rposNext <= data.length) {
+            System.arraycopy(data, rpos, buf, off, len);
+        } else {
+            int tail = data.length - rpos;
+            System.arraycopy(data, rpos, buf, off, tail); // segment right of rpos
+            rposNext = len - tail; // rpos wraps around the end of the buffer
+            System.arraycopy(data, 0, buf, off + tail, rposNext); // remainder
+        }
+        rpos = rposNext;
+    }
+
+    @SuppressWarnings("unchecked")
+    public T putRawBytes(byte[] d, int off, int len) {
```

Comment:
Block comment please. What does this return, and why does it need to return anything at all?